---
title: "Seurat analysis"
linestretch: 0.5
output:
 html_notebook: default
 pdf_document:
   latex_engine: xelatex
---
\fontsize{8}{18}

# Load packages
```{r load-packages, message=F, warning=F}
library(Seurat);library(Matrix)
```

# Import raw count matrix generated by CellRanger
```{r load matrix}
path <- "./outs/raw_gene_bc_matrices/mm10"
counts <- Read10X(path)

# Create some folders to organize the output
# dir.create("figures");dir.create("tables");dir.create("data")
```

# Initialize the Seurat object with the raw count
We keep all genes expressed in >= 3 cells, keep all cells with >= 200 genes.
```{r}
cdS <- new("seurat", raw.data = counts)

cdS <- Setup(cdS, min.cells = 3, min.genes = 200, is.expr = 1, do.logNormalize = T, total.expr = 1e4, project = "Diencephalon")

mito.genes <- grep("^mt-", rownames(cdS@data), value = T)
percent.mito <- colSums(expm1(cdS@data[mito.genes, ])) / colSums(expm1(cdS@data))

#AddMetaData adds columns to object@data.info, and is a great place to stash QC stats
cdS <- AddMetaData(cdS, percent.mito, "percent.mito")
```
# Inspect the quality of cells and genes
```{r fig.width=3, fig.asp=0.3}
VlnPlot(cdS, c("nGene", "nUMI", "percent.mito"), nCol = 3)
```


```{r}
par(mfrow = c(1, 2))
GenePlot(cdS, "nUMI", "percent.mito")
GenePlot(cdS, "nUMI", "nGene")
```

# Filter out outlier cells
We remove out cells that have unique gene counts over 2,500 and include all cells with low counts to ensure glial cells being included. We also remove cells with mitochondrial genes over 5%. 

```{r}
cdS <- SubsetData(cdS, subset.name = "nGene", accept.high = 2500)
cdS <- SubsetData(cdS, subset.name = "percent.mito", accept.high = 0.05)

# Plot the gene counts afer filtering
VlnPlot(cdS, c("nGene", "nUMI", "percent.mito"), nCol = 3)
```

# Regress out unwanted sources of variation
```{r}
# This step take a long time, so we just reload the Seurat object
# cdS <- RegressOut(cdS, latent.vars = c("nUMI", "percent.mito"), model.use = "negbinom")

cdS <- readRDS("./allCells/data/Diencephalon.rds")
```

# Identify variable genes for downstream analysis
```{r}
# Choose variable genes with on mean-variability plot
par(mfrow = c(1, 1))
cdS <- MeanVarPlot(cdS ,fxn.x = expMean, fxn.y = logVarDivMean, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, do.contour = F)
```

```{r}
length(cdS@var.genes)
```

# Perform PCA analysis
```{r}
cdS <- PCAFast(cdS, pc.genes = cdS@var.genes, do.print = TRUE, pcs.print = 20, genes.print = 10,pcs.compute = 40)

PCElbowPlot(cdS, num.pc = 40)
```

# Inspect PCAs with PC heatmap
```{r}
PCHeatmap(cdS, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r}
PCHeatmap(cdS, pc.use = 13:24, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r}
PCHeatmap(cdS, pc.use = 25:36, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

# Run tSNE and find clusters
We skip PC 1 and 3 because they are mostly related to cell cycle.
```{r}
cdS <- RunTSNE(cdS, cells.use = NULL, dims.use = c(2,4:28), k.seed = 1,
               do.fast = T, add.iter = 0, genes.use = NULL,
               reduction.use = "pca", dim_embed = 2, perplexity = 30)

cdS <- FindClusters(cdS, genes.use = NULL, pc.use = c(2,4:28), k.param = 30,
                    k.scale = 25, plot.SNN = FALSE, prune.SNN = 1/10, save.SNN = TRUE,
                    reuse.SNN = FALSE, do.sparse = TRUE, modularity.fxn = 1,
                    resolution = seq(0.5,4,0.5), algorithm = 1, n.start = 100, n.iter = 10,
                    random.seed = 0, print.output = F)
```

# print all tSNE plots with differrent resolutions
```{r}
for (n in seq(0.5,4,0.5)){
  res <- paste("res",n,sep=".")
  cdS <- SetAllIdent(cdS, id = res)
  TSNEPlot(cdS,do.label = T)
  cdS <- BuildClusterTree(cdS, do.reorder = F, reorder.numeric = F, do.plot = F)
}
```


# inspect the clustering with marker genes
```{r}
FeaturePlot(cdS, c("Gbx2","Shh","Ascl1","Lhx1","Pitx2","Olig3","Dbx1","Lhx2","Foxa2","Dlx2","Sox14","Uncx","Hs6st2","Tac1","Islr2","GFP"),
            cols.use = c("lightgrey","red"))
```

# Find appropriate resolution for cell clustering
## Merge clusters
```{r}
# it appears that res=1.5 is appropiate
cdS <- SetAllIdent(cdS, id = "res.1")

#Build a classification hierarchy that places transcriptionally similar clusters adjacent on a tree
cdS <- BuildClusterTree(cdS, do.reorder = T, reorder.numeric = T, do.plot = F)


TSNEPlot(cdS, do.return = T, no.legend = T, do.label = T)
PlotClusterTree(cdS)
```

# Save object
```{r}
# saveRDS(file = "./data/Diencephalon.rds",cdS)
```

# Merge clusters
```{r}
cdS <- RenameIdent(cdS, old.ident.name = 9,new.ident.name = 3)
cdS <- RenameIdent(cdS, old.ident.name = 18,new.ident.name = 17)
```

```{r fig.asp=1}
cdS <- BuildClusterTree(cdS, do.reorder = T, reorder.numeric = T, do.plot = F)

TSNEPlot(cdS, do.return = T, no.legend = T, do.label = T)
PlotClusterTree(cdS)
```
# Save object
```{r}
# saveRDS(file = "./data/Diencephalon_merged.rds",cdS)
# load previous saved object
cdS <- readRDS(file = "./data/Diencephalon_merged.rds")
```

# Examine average gene counts in each cluster
```{r}
FeaturePlot(cdS, c("nGene","nUMI", "percent.mito"), dim.1 = 1, dim.2 = 2, cells.use = NULL,
            pt.size = 1, cols.use = c("blue", "red"), pch.use = 16,
            reduction.use = "tsne", use.imputed = T, nCol = 3,
            no.axes = FALSE)
```

```{r}
VlnPlot(cdS, c("nGene", "nUMI", "percent.mito"), nCol = 1,size.use = 0.1)
```


# Perform differentiation gene expression analysis to find cluster-specific markers

```{r}
library(dplyr)
sig_genes <- FindAllMarkers(cdS, ident.1=NULL, ident.2 = NULL, genes.use = NULL,
                            thresh.use = 0.5, min.diff.pct = 0.01,min.pct = 0.1,
                            test.use = "bimod", 
                            print.bar = T, only.pos = T,
                            max.cells.per.ident = Inf, return.thresh = 0.01, do.print = T)
sig_genes <- mutate(sig_genes,pct.diff=pct.1-pct.2)
write.table(sig_genes,"./tables/sigGenes_all.txt",sep="\t",quote = F)

top <- sig_genes %>% group_by(cluster) %>% top_n(5, avg_diff);top
tab <- as.data.frame(table(cdS@ident))
top$cellNumber <- tab$Freq[match(top$cluster,tab$Var1)]
top$percentage <- round(top$cellNumber/ncol(cdS@data)*100, digits = 2)
write.table(top,"./tables/sigGenes_all_top5.txt",sep="\t",quote = F)
```

# Inspect marker genes
```{r}
sig_genes %>% group_by(cluster) %>% top_n(10, pct.diff) -> top10
cdS.downsampled <- SubsetData(cdS, max.cells.per.ident = 250)
DoHeatmap(cdS.downsampled, genes.use = top10$gene, order.by.ident = TRUE, slim.col.label = TRUE, remove.key = TRUE)
```

```{r}
sig_genes %>% group_by(cluster) %>% top_n(5, pct.diff) -> top5
cdS.downsampled <- SubsetData(cdS, max.cells.per.ident = 250)
DoHeatmap(cdS.downsampled, genes.use = top5$gene, order.by.ident = TRUE, slim.col.label = TRUE, remove.key = TRUE)


sig <- sig_genes %>% arrange(cluster, desc(avg_diff), desc(pct.diff))
top <- sig %>% group_by(cluster) %>% top_n(3);top

cdS.downsampled <- SubsetData(cdS, max.cells.per.ident = 250,random.seed = 1)
DoHeatmap(cdS.downsampled, genes.use = as.character(top$gene), order.by.ident = TRUE, slim.col.label = TRUE, remove.key = TRUE)
```

# Inspect marker gene expression with VlnPlots
## Select post-mitotic cell clusters.
The following steps were run with Seurat version >2.0.
```{r}
cdS <- readRDS(file = "./data/Diencephalon_merged.rds")
cdS@ident <- factor(cdS@ident, ordered = F)
cdS <- UpdateSeuratObject(cdS)
cluster.id <- c(1:7)
cdSsub <- SubsetData(cdS, ident.use = cluster.id, subset.raw =T)
```


```{r}
geneLIst <- c("Gbx2","Lhx2","Cntn2","Islr2")
VlnPlot(cdSsub, features.plot = geneLIst, ident.include = NULL, nCol = 1,
                  do.sort = FALSE, y.max = NULL, same.y.lims = FALSE, size.x.use = 0,
                  size.y.use = 10, size.title.use = 10, adjust.use = 1,
                  point.size.use = -1, cols.use = NULL, group.by = NULL, y.log = FALSE,
                  x.lab.rot = FALSE, y.lab.rot = FALSE, legend.position = "right",
                  single.legend = TRUE, remove.legend = T, do.return = FALSE,
                  return.plotlist = FALSE)
# VlnPlot(cdSsub, features.plot = geneLIst, nCol = 1, ylab.max = 12, do.ret = TRUE,
#   do.sort = FALSE, size.x.use = 16, size.y.use = 16,
#   size.title.use = 20, adjust.use = 1, size.use = 1, cols.use = NULL,
#   group.by = "ident", y.log = F)
```

# Determine the cell cycle score/phase of proliferating cells

```{r}
# Load a list of genes related to cell cycle
cc <- read.csv("/Volumes/jali/Genome/Cell Cycle Genes.csv",stringsAsFactors = F)
id <- unique(cc$mGene[!is.na(cc$mGene)])


cdS1 <- PCA(cdS, pc.genes = id, do.print = TRUE, pcs.print = 1,
               pcs.store = 6, genes.print = 30, use.imputed = FALSE,
               rev.pca = FALSE)

PCHeatmap(cdS1, pc.use = 1:6, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)

# Most of the cell-cycle-related genes are in PC2
PCHeatmap(cdS1, pc.use = 2, num.genes = length(id), cells.use = 1000, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)

# Add cell cycle scores to the metadata
cc.score<- cdS1@pca.rot$PC2
names(cc.score) <- rownames(cdS1@data.info)
cdS1 <- AddMetaData(cdS1, cc.score, "cc.score")

# Check the distribution of the cell cycle score
hist(cc.score)

# Check the distribution of M (Cenpa) and S (Ung) phase cells
FeaturePlot(cdS1, c("Cenpa","Ung"),
            cols.use = c("lightgrey","red"))

# Set -3 and 3 as the cutoff to define cells in S or M phases
cc.class <- rep(0,length(names(cc.score)))
cc.class[cc.score <= -3] <- 1 ## this is G1/S
cc.class[cc.score > 3] <- 2 ## this is G2/M
names(cc.class) <- names(cc.score)

cdS1 <- AddMetaData(cdS1, cc.class, "cc.class")

FeaturePlot(cdS1, "cc.score", dim.1 = 1, dim.2 = 2, cells.use = NULL,
            pt.size = .4, cols.use = c("red", "lightgray","blue"), pch.use = 16,
            reduction.use = "tsne", use.imputed = T, nCol = NULL,
            no.axes = FALSE)

# Plot cells with cell cycle annotation on t-SNE
FeaturePlot(cdS1, "cc.class", dim.1 = 1, dim.2 = 2, cells.use = NULL,
            pt.size = .4, cols.use = c("lightgray","red", "blue"), pch.use = 16,
            reduction.use = "tsne", use.imputed = T, nCol = NULL,
            no.axes = FALSE)
```


# Find genes that are differentiailly expressed in progenitors and postmitotic cells
```{r}
library(Seurat)
cdS <- readRDS("./data/Seurat_allCells.rds")
TSNEPlot(cdS, do.label = T)

cdSsub <- SubsetData(cdS, ident.remove = c(11, 18,17,12), subset.raw = T)

current.cluster.ids <- c(1:7,8:10,13:16)
new.cluster.ids <- c(rep("N",7),rep("NPre",3),rep("NPC",4))
cdSsub@ident <- plyr::mapvalues(x = cdSsub@ident, from = current.cluster.ids, to = new.cluster.ids)
TSNEPlot(cdSsub, do.label = T)

library(dplyr)
mGenes <- readRDS(file="/Volumes/jali/Genome/Annotation1.rds")
sig_genes <- FindAllMarkers(cdSsub)
sig_genes <- mutate(sig_genes,pct.diff=pct.1-pct.2)

sig_genes$TF <- "no"
id <- intersect(mGenes$mgi_symbol,sig_genes$gene)

x <- sig_genes[sig_genes$gene %in% id,]
sig_genes$TF[sig_genes$gene %in% id] <- mGenes$Type[match(x$gene,mGenes$mgi_symbol)]
sig_genes$description <- mGenes$description[match(as.character(sig_genes$gene),mGenes$mgi_symbol)]
head(sig_genes)

sig_genes <- sig_genes[, c("gene","TF","description","p_val","avg_logFC","pct.1","pct.2","pct.diff","cluster")]
openxlsx::write.xlsx(sig_genes,"./tables/sigGenes_NPC1.xlsx")
```

# Show expression of markers
```{r}
FeaturePlot(cdS, c("Id3","Ptn"),overlay = T, cols.use = c("azure2", "red", "blue","green"), no.axes = T) 
FeaturePlot(cdS, c("Stmn2","Gap43"),overlay = T, cols.use = c("azure2", "red", "blue","green"), no.axes = T)
FeaturePlot(cdS, c("Gadd45g","Hes6"),overlay = T, cols.use = c("azure2", "red", "blue","green"), no.axes = T)
```


# Find markers that define different populations of neural progenitor cells
```{r}
cdSsub <- SubsetData(cdS, ident.use = c(12:16), subset.raw = T)
TSNEPlot(cdSsub, do.label = T)

library(dplyr)
mGenes <- readRDS(file="/Volumes/jali/Genome/Annotation1.rds")
sig_genes <- FindAllMarkers(object = cdSsub, logfc.threshold = 0.25,
                            test.use = "wilcox", min.pct = 0.1, min.diff.pct = -Inf,
                            print.bar = TRUE, only.pos = TRUE, max.cells.per.ident = Inf,
                            return.thresh = 0.01, do.print = FALSE, random.seed = 1,
                            min.cells.gene = 3, min.cells.group = 3, latent.vars = "nUMI",
                            assay.type = "RNA")
sig_genes <- mutate(sig_genes,pct.diff=pct.1-pct.2)

sig_genes$TF <- "no"
id <- intersect(mGenes$mgi_symbol,sig_genes$gene)

x <- sig_genes[sig_genes$gene %in% id,]
sig_genes$TF[sig_genes$gene %in% id] <- mGenes$Type[match(x$gene,mGenes$mgi_symbol)]
sig_genes$description <- mGenes$description[match(as.character(sig_genes$gene),mGenes$mgi_symbol)]
head(sig_genes)

sig_genes <- sig_genes[, c("gene","TF","description","p_val","avg_logFC","pct.1","pct.2","pct.diff","cluster")]
openxlsx::write.xlsx(sig_genes,"./tables/sigGenes_NPC.xlsx")
```


# Print session information
```{r}
sessionInfo()
```
